contracts/zaps/ZapTrancheLp.sol

```diff
@@ -0,0 +1,277 @@
+    function _permitCall(PermitData memory data) internal {
+        // Make the permit call to the token in the data field using
+        // the fields provided.
+        // Security note - This fairly open call is safe because it cannot
+        // call 'transferFrom' or other sensitive methods despite the open
+        // scope. Do not make more general without security review.
+        data.tokenContract.permit(
+            msg.sender,
+            data.who,
+            data.amount,
+            data.expiration,
+            data.v,
+            data.r,
+            data.s
+        );
+    }
+
+    struct Out {
+        bytes32 poolId;
+        IVault.ExitPoolRequest request;
+    }
+
+    struct In {
+        uint256 lpCheck;
+        bytes32 poolId;
+        IVault.JoinPoolRequest request;
+    }
+
+    /// @notice Removes PT + YT liquidity from balancer pools, withdraws the PT + YT,
+    /// deposits into a new Tranche, and joins new PT and YT liquidity pools.
+    /// @dev The
+    /// @param _toMint The amount of underlying to mint with in the new tranche.
+    /// @param _ptOutInfo Out struct with info to exit the lp position for the pt.
+    /// @param _ytOutInfo Out struct with info to exit the lp position for the yt.
+    /// @param _ptInInfo In struct with info to enter the new lp position for the pt.
+    /// @param _ytInInfo In struct with info to enter the new lp position for the yt.
+    /// @param _onlyPrincipal This can be set to true to avoid any yt actions.
+    /// @param _permitCallData Encoded array of permit calls to make prior to minting
+    ///                        the data should be encoded with abi.encode(data, "PermitData[]")
+    ///                        each PermitData struct provided will be executed as a call.
+    ///                        An example use of this is if using a token with permit like USDC
+    ///                        to encode a permit which gives this contract allowance before minting.
+    function hopToTranche(
+        uint256 _toMint,
+        Out memory _ptOutInfo,
+        Out memory _ytOutInfo,
+        In memory _ptInInfo,
+        In memory _ytInInfo,
+        bool _onlyPrincipal,
+        PermitData[] calldata _permitCallData
+    ) public notFrozen {
+        IERC20 ytPool;
+        // get tranches
+        ITrancheExt trancheFrom = ITrancheExt(
+            IPool(address(uint160(uint256(_ptOutInfo.poolId)) >> (12 * 8)))
+                .bond()
+        );
+
+        ITrancheExt trancheTo = ITrancheExt(
+            IPool(address(uint160(uint256(_ptInInfo.poolId)) >> (12 * 8)))
+                .bond()
+        );
+
+        // transfer pt lp tokens
+        IERC20 ptPool = IERC20(
+            address(uint160(uint256(_ptOutInfo.poolId)) >> (12 * 8))
+        );
+        ptPool.transferFrom(
+            msg.sender,
+            address(this),
+            ptPool.balanceOf(msg.sender)
+        );
+
+        // exit PT position
+        _vault.exitPool(
```

Comment:
Misconfigured input data can cause some dust locking here, not sure there's a clean fix though.